### PR TITLE
[DO NOT MERGE] [coop] Introduce new runtime handle API

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by kumpera** [comment]
I think this version is a good start. The API itself needs some usability love but we can start from it.

File: mono/metadata/handle.c [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+#endif
+
+		thread->handles_arena = NULL;
+	}
+
+	thread->handles_arena = arena->prev;
+
+	for (chunk = arena->chunk; chunk; chunk = next) {
+		next = chunk->next;
+		memset (&chunk->handles [0], 0, sizeof (sizeof (MonoHandle) * HANDLES_PER_CHUNK));
```
Comment (ludovic-henry) [inline mono/metadata/handle.c]: fix this sizeof (sizeof (...))

File: mono/metadata/handle.c [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+	thread = mono_thread_internal_current ();
+	g_assert (thread);
+
+	if (!arena->prev) {
+#ifdef HAVE_SGEN_GC
+		mono_gc_deregister_root ((char*) &thread->handles_arena);
+#endif
+
+		thread->handles_arena = NULL;
```
Comment (ludovic-henry) [inline mono/metadata/handle.c]: remove that set

**Comment by ludovic-henry** [comment]
@schani I will work on adding a free list to avoid allocating new HandleArena and HandleArenaChunk every time. The number of handles should still, most of the time be pretty small, because you need only one per object, and there is few icalls that receive and returns a lot of them.

**Comment by ludovic-henry** [comment]
depends on https://github.com/mono/mono/pull/2231

**Comment by alexischr** [comment]
Are these handles only for recording managed references in native code? If so, I think it can be simpler if the arena is a TLAB a la SGen:

* TLAB is lazily allocated on the first MONO_HANDLE_ARENA_START for thread. Static size should be fine
* allocation by bumping the pointer
* pointer resets to start on MONO_HANDLE_ARENA_END.

Changes to use the handle look (unavoidably) clunky... We can consider recording the references   before they are returned to us by the runtime functions (e.g. if thread is in a "record my references" mode, the refs returned are added to the arena)

(But that would mean we have to pin the object, ->obj does have the advantage of being moveable...)

**Comment by ludovic-henry** [comment]
@alexischr if I understand TLAB correctly, these handles are allocated on a TLAB (the arena is attached to a thread), and they are freed when calling MONO_HANDLE_ARENA_END. The only difference I see from your suggestion is that you suggest allocating what I call a chunk (a piece of memory containing the Handles) statically vs dynamically (how I do it for now).
As an optimization, we could definitely have the first chunk being part of the MonoHandleArena, so if you have only a few handles allocated (less than 14), then we only have 1 allocation vs 2 right now.

I will resplit the PR to remove the dependency on #2231 as that just make it unreadable for no added value.

**Comment by alexischr** [comment]
Ah now I see you added to MonoInternalThread, though in that case you don't need the InterlockedExchange ops (or the memset)

I wouldn't worry about chunks and chaining either right now, we can give it a few kilobytes and watch some results first.

**Pull request closed without merging**